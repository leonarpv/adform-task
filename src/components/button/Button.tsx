import React, { ButtonHTMLAttributes, DetailedHTMLProps } from 'react'
import classNames from 'classnames'
import './button.scss'

type Props = DetailedHTMLProps<
    ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
> & {
    secondary?: boolean
}

const Button = ({ secondary, ...rest }: Props) => {
    const className = classNames('button', {
        'button--secondary': secondary,
    })

    return <button className={className} {...rest} />
}

export default Button
