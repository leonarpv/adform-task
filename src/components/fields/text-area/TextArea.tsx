import React, { DetailedHTMLProps, TextareaHTMLAttributes } from 'react'
import BaseField, { BaseFieldProps } from '../BaseField'
import './textArea.scss'

type Props = DetailedHTMLProps<
    TextareaHTMLAttributes<HTMLTextAreaElement>,
    HTMLTextAreaElement
> &
    BaseFieldProps

const TextArea = ({ label, error, ...rest }: Props) => {
    return (
        <BaseField label={label} error={error}>
            <textarea className="text-area" {...rest} />
        </BaseField>
    )
}

export default TextArea
