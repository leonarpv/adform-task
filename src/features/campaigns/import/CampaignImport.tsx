import React, { ChangeEvent, FormEvent, useState } from 'react'
import Modal from 'react-modal'
import Button from 'components/button/Button'
import TextArea from 'components/fields/text-area/TextArea'
import './campaignImport.scss'

type Props = {
    onImport: (input: string) => void
}

const CampaignImport = ({ onImport }: Props) => {
    const [isOpen, setIsOpen] = useState<boolean>(false)
    const [value, setValue] = useState<string>('')

    const openModal = () => {
        setIsOpen(true)
    }

    const closeModal = () => {
        setIsOpen(false)
    }

    const handleChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
        setValue(e.target.value)
    }

    const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault()

        onImport(value)
    }

    return (
        <>
            <Button onClick={openModal}>Import Campaigns</Button>
            <Modal
                appElement={document.body}
                isOpen={isOpen}
                onRequestClose={closeModal}
                contentLabel="Example Modal"
            >
                <form className="campaign-import" onSubmit={handleSubmit}>
                    <TextArea
                        label="Enter Campaign Data"
                        name="message"
                        value={value}
                        onChange={handleChange}
                    />
                    <div className="campaign-import__actions">
                        <Button type="submit">Import</Button>
                        <Button type="button" onClick={closeModal} secondary>
                            Cancel
                        </Button>
                    </div>
                </form>
            </Modal>
        </>
    )
}

export default CampaignImport
