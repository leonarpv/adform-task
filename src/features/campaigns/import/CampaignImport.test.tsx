import { shallow } from 'enzyme'
import React, { ChangeEvent, FormEvent } from 'react'
import Modal from 'react-modal'
import CampaignImport from './CampaignImport'
import Button from 'components/button/Button'
import TextArea from 'components/fields/text-area/TextArea'
import { Mock } from 'ts-mockery'

describe('CampaignImport', () => {
    it('should render properly', () => {
        const onImport = jest.fn()

        const wrapper = shallow(<CampaignImport onImport={onImport} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('should render datepickers and input', () => {
        const onImport = jest.fn()

        const wrapper = shallow(<CampaignImport onImport={onImport} />)

        expect(wrapper.find(Button).length).toBe(3)
        expect(wrapper.find(Modal).exists()).toBe(true)
        expect(wrapper.find(TextArea).exists()).toBe(true)
        expect(wrapper.find('form').exists()).toBe(true)
    })

    it('should call onImport on TextArea change', () => {
        const onImport = jest.fn()

        const wrapper = shallow(<CampaignImport onImport={onImport} />)

        const submitEvent = Mock.of<FormEvent<HTMLFormElement>>({
            preventDefault: jest.fn(),
        })

        const changeEvent = Mock.of<ChangeEvent<HTMLTextAreaElement>>({
            target: {
                value: 'value',
            },
        })

        const onInputChange = wrapper.find(TextArea).props().onChange

        expect(onInputChange).toBeDefined()

        onInputChange && onInputChange(changeEvent)

        const onSubmit = wrapper.find('form').props().onSubmit

        expect(onSubmit).toBeDefined()

        onSubmit && onSubmit(submitEvent)

        expect(onImport).toHaveBeenCalledTimes(1)
        expect(onImport).toHaveBeenCalledWith(changeEvent.target.value)
    })
})
