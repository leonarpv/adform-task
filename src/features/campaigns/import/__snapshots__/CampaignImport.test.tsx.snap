// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignImport should render properly 1`] = `ShallowWrapper {}`;
